Add TopN optimization in physical plan mapping

kryonix opened the pull request:
This PR fixes #11260 by introducing the TopN optimization to the physical plan mapping of `LogicalLimit` operators.

kryonix committed:
Add TopN optimization in physical plan mapping

Fixes #11260

kryonix committed:
Apply TopN optimization only if no projection is used

kryonix commented:
This should now work as intended, _i.e._, apply the TopN optimization when possible during physical plan mapping.
The principle idea of this PR is that unnecessary projections are eliminated when the physical plan is created. This then allows us to use the TopN operator instead of LIMIT and ORDER BY, when certain conditions are met.

kryonix committed:
Fix format

kryonix committed:
Fix tidy check

Tmonster reviewed (changes requested):
Thanks for the PR!

This will definitely work, however, we like to avoid having these kind of optimizations during the physical planning phase whenever possible. If I'm not mistaken we can move this logic to the the topn optimizer itself. It would require just a bit more work on the PR. If the operator is a limit, you can loop through any child projections until you get to an order by. If you encounter a different operator before then, you don't apply the optimization.

If you could add a test and a micro benchmark that would be great!
See `test/optimizer/topn_optimizer.test` for test examples, 
`benchmark/micro/limit/parallel_limit.benchmark` is a minor benchmark example you can copy and modify. Maybe a different directory though, `benchmark/micro/optimizer/` would be good.

kryonix commented:
Thanks for the feedback @Tmonster, I will take a look at your suggestions 👍🏻

kryonix committed:
Improve TopN optimization pass

Remove TopN optimization from physical plan mapping

Add TopN optimization test

Add TopN optimization benchmark

kryonix committed:
Revert changes in plan_limit.cpp

kryonix committed:
Fix topN optimizer test

kryonix committed:
Fix generated check

Tmonster reviewed (changes requested):
Thanks for the fixes!

Looks much better now. Still some improvements to be made. Let me know if you need any help

Tmonster commented:
Also can you rebase with feature and change the base of this PR to merge to feature? 
We are working on release v1.0.0 so only bug fixes are getting merged now. Optimizers are a gray area, but this is probably a feature instead of a bug

kryonix committed:
Improve topN with projection pullup optimization

kryonix commented on src/optimizer/topn_optimizer.cpp:
```diff
@@ -38,6 +38,66 @@ unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
+
+				// Check if there are only projection operators between the limit operator
+				// and an order by operator.
+				while (child_op && !can_optimize) {
```
Makes sense! I fixed that.

kryonix commented on test/optimizer/topn_optimizer.test:
```diff
@@ -37,3 +37,9 @@ query II
+
+# only limit, but in subquery with projection pull up optimization
+query II
+EXPLAIN SELECT * FROM (SELECT * FROM range(100000000) AS _(x) ORDER BY x) AS cte LIMIT 10
```
Good catch. I've added the test.

kryonix commented on src/optimizer/topn_optimizer.cpp:
```diff
@@ -38,6 +38,66 @@ unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
+				if (can_optimize) {
+					// traverse operator tree and collect all projection nodes until we reach
+					// the order by operator
+					vector<unique_ptr<LogicalOperator>> projections;
```
This code no longer exists separately. I've merged it with the existing case.

kryonix commented:
Thank you for the review @Tmonster, I've added all the changes. I did rebase to feature, too.

Tmonster commented on src/optimizer/topn_optimizer.cpp:
```diff
@@ -20,13 +19,41 @@ bool TopN::CanOptimize(LogicalOperator &op) {
+
+		vector<unique_ptr<LogicalOperator>> projections;
+
+		if (op->children[0]->type != LogicalOperatorType::LOGICAL_ORDER_BY) {
```
something like might look nicer. If there is no projection sitting in the way, the while loop won't run. You already know you can optimize, so the cast will be valid (and the `D_ASSERT` will check it in debug mode.
```
auto child = std::move(op->children[0]);
// collect all projections until we get to the order by
while (child->type == LogicalOperatorType::LOGICAL_PROJECTION) {
	D_ASSERT(!child->children.empty());
	auto tmp = std::move(child->children[0]);
	projections.push_back(std::move(child));
	child = std::move(tmp);
}
D_ASSERT(child->type == LogicalOperatorType::LOGICAL_ORDER_BY);
auto &order_by = child->Cast<LogicalOrder>();
```
might look nicer.

kryonix replied:
Awesome, that looks good to me.

Tmonster reviewed (changes requested):
Nice! Looks very good now, just 2 more small comments then it's good to go 👍

kryonix committed:
Remove unnecessary if and else branch

kryonix commented on src/optimizer/topn_optimizer.cpp:
```diff
@@ -37,6 +64,14 @@ unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
+			projections.pop_back();
+		}
 	} else {
 		for (auto &child : op->children) {
```
Good catch. I did remove that.

Tmonster reviewed (approved):
Looks good to me! Thanks 👍 
Our CI to automatically run tests is a bit broken right now. So to run CI, you need to manually convert it to a draft, then convert it back to say ready to review. You can find the button to it it under the "reviewers"

Tmonster commented:
Hi @kryonix ,

I think you need to run a `make format-fix` and the Code quality CI will turn green. I don't know why the Linux extensions CI is failing. I think that is just noise.

kryonix committed:
Merge branch 'feature' of github.com:cwida/duckdb into topn_11260

kryonix commented:
Hi @Tmonster, I did just that before my last commit. `make format-fix` does not make a difference. I merged with feature just now because there were some file conflicts. Maybe that will fix it.

Tmonster reviewed (approved):
LGTM!
